### ISSUE TITLE
boot.initrd.network.wireguard options

### ISSUE THREAD

**Issue by genofire** [NixOS/nixpkgs#141632]
**Project description**
It would be nice to have an vpn connection on stage1 to open crypted HDDs.
openvpn is there already, i prefer wireguard

**Metadata**

* homepage URL:
* source URL:
* license: mit, bsd, gpl2+ , ...
* platforms: unix, linux, darwin, ...

**Comment by genofire** [comment]
No reaction, but still interresting for me.

**Comment by matthiasbeyer** [comment]
I would be interested in this as well.

**Comment by BeatLink** [comment]
I ended up using the boot.initrd.systemd.network options to get it done

**Comment by 456634** [comment]
@BeatLink can you share the part of your config pls? :)

**Comment by BeatLink** [comment]
The following setup allows for remote LUKS decryption over WireGuard. It should act as a good starting point



    # Wireguard Remote LUKS Unlock ------------------------------------------------------------------------------------------------------------------
    boot.initrd = {
        network = {
            ssh = { # Enables SSH in the initrd for remote LUKS unlocking 
                enable = true;
                authorizedKeys = ["<Authorized SSH Keys for Remote Login and Decryption>"];
                hostKeys = ["<SSH Host Key>"];
            };
        };
        systemd = {
            enable = true;
            contents = { # Make the private key accessible in the systemd initrd
                "/wireguard-initrd-key".source = /wireguard-initrd-privatekey;
            };
            # Upon SSH login, call the LUKS unlock function
            users.root.shell = "/bin/systemd-tty-ask-password-agent";
            network = { # Enable systemd networking in the initrd
                enable = true;
                netdevs."255-wg255" = { # Create the wireguard network device
                    netdevConfig = {
                        Kind = "wireguard";
                        Name = "wg255";
                        MTUBytes = "1300";
                    };
                    wireguardConfig = {
                        PrivateKeyFile = "/wireguard-initrd-key";
                        ListenPort = 44444;
                    };
                    wireguardPeers = [
                        {
                            # Laptop
                            wireguardPeerConfig = {
                                PublicKey = "<Public Key of Peer>";
                                AllowedIPs = ["<IP Address of Peer>"];
                            };
                        }
                    ];
                };
                networks = {  # Setup the Ethernet Network
                    "255-lan" = {
                        matchConfig.Name = "enp2s0f1";
                        address = [ "192.168.0.2/24"];
                        routes = [{ routeConfig.Gateway = "192.168.0.1"; }];
                        linkConfig.RequiredForOnline = "routable";
                    };
                    "255-wireguard" = { # Setup the Wireguard Network
                        matchConfig.Name = "wg255";
                        address = ["10.100.100.254/24"];
                    };
                };
                
            };
        };
    };

**Comment by BeatLink** [comment]
Yeah i've encountered the same issues. What i ended up doing was performing all of my network configuration in initramfs and just keep the configuration when the OS is booted

**Comment by almereyda** [comment]
Is the behaviour maybe related to these issues?

- https://github.com/NixOS/nixpkgs/issues/186963
- https://github.com/NixOS/nixpkgs/issues/350531

**Comment by BeatLink** [comment]
An updated solution ive found was to configure all of the networking in the main systemd-networkd environment then simply use the exact same settings for the initrd network. Systemd-Networkd will then use the same devices and interface in both environments without conflicting with each other. 


Here's the abridged config for my backup server

```nix
{
    config,
    inputs,
    ...
}:
{
    networking = {
        # Use Systemd-Networkd
        useNetworkd = true;

        # Sets the Wireguard interface as trusted in the firewall
        firewall.trustedInterfaces = [ "wg0" ];
    };

    # Loads the Wireguard private key from SOPS and sets the permissions to systemd-networkd
    sops.secrets.wireguard_private_key = {
        sopsFile = "${inputs.self}/secrets/1-backup-server.yaml";
        owner = "systemd-network";
        group = "systemd-network";
    };

    boot.initrd = {
        # Needed for wireguard in initrd for remote unlocking
        availableKernelModules = [ "wireguard" ];

        # Sops doesn't work in initrd so we use boot.initrd.secrets
        secrets."${config.sops.secrets.wireguard_private_key.path}" =
            config.sops.secrets.wireguard_private_key.path;

        systemd = {
            # The Wireguard privatekey must be owned by systemd-network to be used.
            services = {
                fix_wireguard_key_perms = {
                    description = "Set permissions for wireguard private key";
                    wantedBy = [ "initrd.target" ];
                    after = [ "initrd-nixos-copy-secrets.service" ];
                    before = [ "systemd-networkd.service" ];
                    unitConfig.DefaultDependencies = "no";
                    serviceConfig.Type = "oneshot";
                    script = ''chown systemd-network:systemd-network "${config.sops.secrets.wireguard_private_key.path}" '';
                };
            };

            # Sets up systemd-networkd in initrd using the same configuration from the booted system's network stack
            network = config.systemd.network;
        };
    };

    systemd = {
        network = {
            enable = true;
            netdevs."wg0" = {
                netdevConfig = {
                    Kind = "wireguard";
                    Name = "wg0";
                };
                wireguardConfig = {
                    PrivateKeyFile = config.sops.secrets.wireguard_private_key.path;
                    ListenPort = 51820;
                };
                wireguardPeers = [
                    {
                        # Server
                        PublicKey = "<Wireguard Peer Public Key>";
                        AllowedIPs = [ "10.100.100.0/24" ];
                        Endpoint = "<Peer Endpoint IP:Port>";
                        PersistentKeepalive = 25;
                    }
                ];
            };
            networks = {
                # My ethernet interface
                "end0" = {
                    matchConfig.Name = "end0";
                    networkConfig.DHCP = "ipv4";
                    linkConfig.RequiredForOnline = "routable";
                };
                "wg0" = {
                    matchConfig.Name = "wg0";
                    address = [ "10.100.100.5/24" ];
                    dns = [ "10.100.100.1" ];
                };
            };
        };
    };
}

```

**Comment by BeatLink** [comment]
A bonus with the above solution is that there's no need to create a seperate wireguard interface for just the initrd. you can use the same one from the booted system.

**Comment by almereyda** [comment]
Given that, in some cases it might still be desireable to drop an initrd WireGuard interface, such as with trusted networks that contain a Tang server which is only needed in initrd by Clevis.
